apps/vmq_ql/src/vmq_ql_query.erl

```diff
@@ -0,0 +1,380 @@
+collect_results(_, 0, NextKey, _, Acc) -> {NextKey, lists:reverse(lists:flatten(Acc))};
+collect_results(desc, Limit, Key, Tab, Acc) ->
+    collect_results(desc, Limit - 1, ets:next(Tab, Key), Tab,
+                    [ets:lookup(Tab, Key)|Acc]);
+collect_results(asc, Limit, Key, Tab, Acc) ->
+    collect_results(asc, Limit - 1, ets:prev(Tab, Key), Tab,
+                    [ets:lookup(Tab, Key)|Acc]).
+
+cleanup_result(#state{result_table=undefined} = State) -> State;
+cleanup_result(#state{result_table=Tab} = State) ->
+    ets:delete(Tab),
+    State#state{result_table=undefined}.
+
+internal_query(Str) ->
+    try vmq_ql_parser:parse(Str) of
+        {fail, E} ->
+            {error, E};
+        Parsed when is_list(Parsed) ->
+            eval(proplists:get_value(type, Parsed), Parsed)
+    catch
+        E:R ->
+            {error, {E, R}}
+    end.
+
+eval("SELECT", Query) ->
+    From = proplists:get_value(from, Query),
+    Fields = proplists:get_value(fields, Query),
+    Where = proplists:get_value(where, Query),
+    OrderBy = proplists:get_value(orderby, Query),
+    Limit = proplists:get_value(limit, Query),
+    select(Fields, module(From), Where, OrderBy, Limit).
+
+select(Fields, Module, Where, OrderBy, Limit) ->
+    FieldsConfig = Module:fields_config(),
+    RequiredFields = lists:usort((required_fields(Where)
+                                  ++ include_fields(FieldsConfig, Fields)
+                                  ++ OrderBy)) -- [all],
+    RowInitializer = get_row_initializer(FieldsConfig, RequiredFields),
+    EmptyResultRow = empty_result_row(FieldsConfig, Fields),
+    Results = ets:new(?MODULE, [ordered_set]),
+    try
+        Module:fold_init_rows(
+          fun(InitRow, Idx) ->
+                  PreparedRows = initialize_row(RowInitializer, InitRow),
+                  lists:foldl(
+                    fun(Row, AccIdx) ->
+                            put({?MODULE, row_data}, Row),
+                            case eval_query(Where) of
+                                true ->
+                                    raise_enough_data(AccIdx, OrderBy, Limit, enough_data),
+                                    Key = order_by_key(AccIdx, OrderBy, Row),
+                                    ets:insert(Results,
+                                               {Key, filter_row(Fields, EmptyResultRow, Row)}),
+                                    AccIdx + 1;
+                                false -> AccIdx
+                            end
+                    end, Idx, PreparedRows)
+          end, 1)
+    catch
+        exit:enough_data ->
+            %% Raising inside an ets:fold allows us to stop the fold
+            ok;
+        E1:R1 ->
+            ets:delete(Results),
+            lager:error("Select query terminated due to ~p ~p", [E1, R1]),
+            exit({E1, R1})
+    end,
+    case is_integer(Limit) of
+        false ->
+            ok;
```

Comment:
How can `Limit` not be an integer at this point? If it is supposed to only be an integer maybe put a guard on the `vmq_ql_query:fetch/3` function as well on `vmq_ql_query_mgr:fold/4` function and skip the check here?